tests/kernels/test_prefix_prefill.py

```diff
@@ -207,3 +207,245 @@ def test_contexted_kv_attention(
+            end_slot = start_slot + end_loc - start_loc
+            k_cache.view(-1, num_kv_heads,
+                         head_size)[start_slot:end_slot].copy_(
+                             key[start_loc:end_loc])
+            v_cache.view(-1, num_kv_heads,
+                         head_size)[start_slot:end_slot].copy_(
+                             value[start_loc:end_loc])
+            cur_ctx += block_size
+            block_id += 1
+    # transpose K_cache[num_blocks, block_size, num_kv_heads, head_size]
+    # to K_cache[num_blocks, num_kv_heads, head_size/8, block_size, 8]
+    k_cache = k_cache.view(-1, block_size, num_kv_heads, head_size // 8,
+                           8).permute(0, 2, 3, 1, 4).contiguous()
+    # transpose V_cache[num_blocks, block_size, num_kv_heads, head_size]
+    # to V_cache[num_blocks, num_kv_heads, head_size, block_size]
+    v_cache = v_cache.view(-1, block_size, num_kv_heads,
+                           head_size).permute(0, 2, 3, 1).contiguous()
+
+    # Warm up the Triton kernel by calling it once before actually measuring
+    # generation time
+    context_attention_fwd(query,
+                          k,
+                          v,
+                          output,
+                          k_cache,
+                          v_cache,
+                          block_table,
+                          b_start_loc,
+                          b_seq_len,
+                          b_ctx_len,
+                          max_input_len,
+                          alibi_slopes=alibi_slopes)
+    torch.cuda.synchronize()
+    start_time = time.time()
+    context_attention_fwd(query,
+                          k,
+                          v,
+                          output,
+                          k_cache,
+                          v_cache,
+                          block_table,
+                          b_start_loc,
+                          b_seq_len,
+                          b_ctx_len,
+                          max_input_len,
+                          alibi_slopes=alibi_slopes)
+    torch.cuda.synchronize()
+    end_time = time.time()
+    print(f"triton Time: {(end_time - start_time)*1000:.2f} ms")
+    scale = float(1.0 / (head_size**0.5))
+
+    # NOTE(DefTruth): In order to reuse _make_alibi_bias function,
+    # we have to pad query tensor before MQA/GQA expanding.
+    if query.shape[0] != key.shape[0]:
+        query_pad = torch.empty(sum(seq_lens),
+                                num_heads,
+                                head_size,
+                                dtype=dtype)
+        query_pad.uniform_(-1e-3, 1e-3)
+        seq_start = 0
+        query_start = 0
+        for i, (query_len, seq_len) in enumerate(zip(query_lens, seq_lens)):
+            seq_end = seq_start + seq_len
+            query_end = query_start + query_len
+            query_pad[seq_start:seq_end, ...] = torch.cat([
+                torch.zeros(
```

Comment:
isn't this supposed to be opposite (to pad to the end)? (torch.cat([query[...], torch.zeros(..)]))?

Reply:
@rkooo567 Because we have assumed that query_lens here represents new tokens that are not hit by the prefix cache. So, it seems more reasonable that old tokens (meaning the kv cache of these old tokens already in the prefix cache) should be padded to the left. This is also consistent with the assumption that the attention mask goes from left to right. I haven't tested padding it to the right as that doesn't seem reasonable.

Reply:
ah I think that makes sense! Thanks for the explanation! can you just quickly check if padding to the beginning fails the test?

Reply:
@rkooo567 yes, the tests will failed because of `AssertError `after changed the padding side to right.   
```bash
test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-128-8-64] triton Time: 8.21 ms
xformers Time: 33.51 ms
FAILED
test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-128-64-64] triton Time: 8.14 ms
xformers Time: 11.26 ms
FAILED
test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-96-1-64] triton Time: 8.60 ms
xformers Time: 8.30 ms
FAILED
test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-96-8-64] triton Time: 7.87 ms
xformers Time: 30.91 ms
FAILED
test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-96-64-64] triton Time: 7.70 ms
xformers Time: 11.08 ms
FAILED
test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-80-1-64] triton Time: 8.10 ms
xformers Time: 8.38 ms
FAILED
test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-80-8-64] triton Time: 7.40 ms
xformers Time: 32.75 ms
FAILED
test_prefix_prefill.py:458: AssertionError
================================================================================ short test summary info =================================================================================
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-128-1-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-128-8-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-128-64-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-96-1-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-96-8-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-96-64-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-80-1-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-80-8-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-80-64-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-24-1-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-24-8-64] - AssertionError: assert False
FAILED test_prefix_prefill.py::test_contexted_kv_attention_alibi[cuda:0-dtype0-24-64-64] - AssertionError: assert False
================================================================================== 12 failed in 42.69s ===================================================================================
```
we have used LowerTriangularMaskWithTensorBias inside _make_alibi_bias, so, i think the padding side of query tensor can not change to right.